File: AgeOfArenaUnity/Assets/Scripts/ScenarioEditor.cs
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

/// <summary>
/// N12.edit: Runtime scenario/map editor.
/// Toggle with 'E' key (or HUD pause-menu button). An overlay canvas appears over the
/// arena; the player can place/delete units, buildings, and resource nodes, configure
/// per-team starting resources/civ, author triggers, then save a scenario and playtest.
///
/// Placement flow: click entity in the left palette → click on the terrain → entity spawns.
/// Delete flow: enable delete mode (🗑 button) → click any entity on the map.
/// Playtest flow: "Test Et" serialises the current scene state and restarts via PendingLoad.
/// Save/load: scenario stored in PlayerPrefs as JSON (key "AoA_Scenario_0").
/// </summary>
public class ScenarioEditor : MonoBehaviour
{
    const string ScenarioKey = "AoA_Scenario_0";

    static readonly Color PanelBg  = new Color(0.06f, 0.08f, 0.14f, 0.92f);
    static readonly Color BtnColor = new Color(0.14f, 0.20f, 0.34f, 1f);
    static readonly Color BtnSel   = new Color(0.22f, 0.42f, 0.68f, 1f);
    static readonly Color BtnDel   = new Color(0.45f, 0.10f, 0.10f, 1f);
    static readonly Color TextCol  = new Color(0.92f, 0.94f, 1.00f, 1f);
    static readonly Color LabelCol = new Color(0.70f, 0.76f, 0.60f, 1f);

    // ── State ─────────────────────────────────────────────────────────────────

    bool _open;
    bool _deleteMode;
    int  _selectedTeam;    // team to assign to newly placed entities
    int  _selectedCategory; // 0=Units 1=Buildings 2=Resources

    static readonly string[] Categories = { "Birimler", "Binalar", "Kaynaklar" };

    // palette items (display-name, action-key)
    static readonly (string label, System.Action<Vector3, Color> spawn)[] UnitItems =
    {
        ("Köylü",   (p,c) => Reg(UnitFactory.Villager(Root(), p, c))),
        ("Piyade",  (p,c) => Reg(UnitFactory.Militia(Root(), p, c))),
        ("Okçu",    (p,c) => Reg(UnitFactory.Archer(Root(), p, c))),
        ("Süvari",  (p,c) => Reg(UnitFactory.Cavalry(Root(), p, c))),
        ("Kargıcı", (p,c) => Reg(UnitFactory.Spearman(Root(), p, c))),
        ("Trebuchet",(p,c) => Reg(UnitFactory.Trebuchet(Root(), p, c))),
        ("Ram",     (p,c) => Reg(UnitFactory.Ram(Root(), p, c))),
        ("Mangonel",(p,c) => Reg(UnitFactory.Mangonel(Root(), p, c))),
    };

    static readonly (string label, System.Action<Vector3, Color, int> spawn)[] BuildingItems =
    {
        ("Şehir Merkezi", (p,c,t) => RegBGo(BuildingFactory.TownCenter(Root(), p, c), t)),
        ("Kışla",         (p,c,t) => RegBGo(BuildingFactory.Barracks(Root(), p, c), t)),
        ("Maden Kampı",   (p,c,t) => RegBGo(BuildingFactory.MiningCamp(Root(), p, c), t)),
        ("Ev",            (p,c,t) => RegBGo(BuildingFactory.House(Root(), p, c), t)),
        ("Kule",          (p,c,t) => RegBGo(BuildingFactory.WatchTower(Root(), p, c), t)),
        ("Kale",          (p,c,t) => RegBGo(BuildingFactory.Castle(Root(), p, c), t)),
    };

    static readonly (string label, System.Action<Vector3> spawn)[] ResourceItems =
    {
        ("Altın Madeni",  p => RegN(ResourceFactory.GoldMine(Root(), p))),
        ("Taş Madeni",    p => RegN(ResourceFactory.StoneMine(Root(), p))),
        ("Orman Ağacı",   p => RegN(ResourceFactory.Tree(Root(), p, ResourceFactory.TreeKind.Broadleaf))),
        ("Böğürtlen",     p => RegN(ResourceFactory.BerryBush(Root(), p))),
        ("Balık Göleti",  p => RegN(ResourceFactory.FishPond(Root(), p))),
    };

    int  _selectedItem = -1;
    Text _statusText;
    Canvas _canvas;

    // Button refs for selection highlight
    readonly List<(Button btn, int idx)> _paletteButtons = new();

    // ── Registration helpers ──────────────────────────────────────────────────

    static Transform Root()
    {
        var gm = GameManager.Instance;
        return gm != null ? gm.transform : null;
    }

    static void Reg(UnitEntity e)
    {
        if (e == null) return;
        GameManager.Instance?.RegisterUnit(e);
    }

    static void RegB(BuildingEntity b)
    {
        if (b == null) return;
        GameManager.Instance?.RegisterBuilding(b);
    }

    // BuildingFactory returns GameObject; extract BuildingEntity and set team.
    static void RegBGo(GameObject go, int teamId)
    {
        if (go == null) return;
        var b = go.GetComponent<BuildingEntity>();
        if (b == null) return;
        b.teamId = teamId;
        GameManager.Instance?.RegisterBuilding(b);
    }

    static void RegN(ResourceNode n)
    {
        if (n == null) return;
        GameManager.Instance?.RegisterNode(n);
    }

    // ── Toggle ────────────────────────────────────────────────────────────────

    public bool IsOpen => _open;

    public void Open()
    {
        _open = true;
        BuildUI();
        _canvas.gameObject.SetActive(true);
        Time.timeScale = 0f; // freeze while editing
    }

    public void Close()
    {
        _open = false;
        if (_canvas != null) _canvas.gameObject.SetActive(false);
        Time.timeScale = 1f;
    }

    void Update()
    {
        if (Input.GetKeyDown(KeyCode.E) && !_open) Open();
        if (!_open) return;

        if (Input.GetKeyDown(KeyCode.Escape)) { Close(); return; }

        // Mouse placement on Left-Click
        if (Input.GetMouseButtonDown(0) && _selectedItem >= 0)
        {
            if (TryGetGroundPoint(out var pos))
            {
                PlaceSelected(pos);
            }
        }
        // Delete on Left-Click in delete mode
        else if (Input.GetMouseButtonDown(0) && _deleteMode)
        {
            TryDeleteAt();
        }
    }

    // ── Placement & delete ────────────────────────────────────────────────────

    void PlaceSelected(Vector3 pos)
    {
        var gm = GameManager.Instance;
        if (gm == null) return;
        var col = TeamPalette.For(_selectedTeam);
        pos.y = 0f;

        if (_selectedCategory == 0 && _selectedItem < UnitItems.Length)
        {
            UnitItems[_selectedItem].spawn(pos, col);
            // Most factories set teamId=0 by default; fix to selected team.
            if (gm.units.Count > 0)
            {
                var u = gm.units[gm.units.Count - 1];
                if (u != null) { u.teamId = _selectedTeam; }
            }
        }
        else if (_selectedCategory == 1 && _selectedItem < BuildingItems.Length)
        {
            BuildingItems[_selectedItem].spawn(pos, col, _selectedTeam);
        }
        else if (_selectedCategory == 2 && _selectedItem < ResourceItems.Length)
        {
            ResourceItems[_selectedItem].spawn(pos);
        }
        SetStatus($"Yerleştirildi: {GetItemName(_selectedCategory, _selectedItem)} (T{_selectedTeam})");
    }

    void TryDeleteAt()
    {
        if (!TryGetWorldHit(out var hit)) return;
        var gm = GameManager.Instance;
        if (gm == null) return;
        var go = hit.collider.gameObject;

        var unit = go.GetComponentInParent<UnitEntity>();
        if (unit != null) { gm.units.Remove(unit); Object.Destroy(unit.gameObject); SetStatus("Birim silindi."); return; }

        var bldg = go.GetComponentInParent<BuildingEntity>();
        if (bldg != null) { gm.buildings.Remove(bldg); Object.Destroy(bldg.gameObject); SetStatus("Bina silindi."); return; }

        var node = go.GetComponentInParent<ResourceNode>();
        if (node != null) { gm.nodes.Remove(node); Object.Destroy(node.gameObject); SetStatus("Kaynak silindi."); return; }
    }

    static bool TryGetGroundPoint(out Vector3 pos)
    {
        pos = Vector3.zero;
        var cam = Camera.main;
        if (cam == null) return false;
        var ray = cam.ScreenPointToRay(Input.mousePosition);
        var plane = new Plane(Vector3.up, Vector3.zero);
        if (!plane.Raycast(ray, out float enter)) return false;
        pos = ray.GetPoint(enter);
        return true;
    }

    static bool TryGetWorldHit(out RaycastHit hit)
    {
        hit = default;
        var cam = Camera.main;
        if (cam == null) return false;
        var ray = cam.ScreenPointToRay(Input.mousePosition);
        return Physics.Raycast(ray, out hit, 200f);
    }

    // ── Save / Load / Playtest ────────────────────────────────────────────────

    void SaveScenario()
    {
        var gm = GameManager.Instance;
        if (gm == null) return;
        var wr = Object.FindAnyObjectByType<WorldRoot>();
        var data = new SaveSystem.SaveData { mapSeed = wr != null ? wr.mapSeed : 0 };

        data.teams = new SaveSystem.TeamSnap[gm.TeamCount];
        for (int t = 0; t < gm.TeamCount; t++)
        {
            var r = gm.teamRes[t];
            data.teams[t] = new SaveSystem.TeamSnap
            {
                food = r.food, wood = r.wood, gold = r.gold, stone = r.stone,
                ageIndex = (int)gm.teamTech[t].age,
                civ = (int)gm.teamCivs[t],
            };
            foreach (TechType ty in System.Enum.GetValues(typeof(TechType)))
                if (gm.teamTech[t].Has(ty)) data.teams[t].techs.Add((int)ty);
        }
        foreach (var u in gm.units)
        {
            if (u == null) continue;
            data.units.Add(new SaveSystem.UnitSnap {
                type = (int)u.type, teamId = u.teamId,
                x = u.transform.position.x, z = u.transform.position.z, hp = u.maxHp,
            });
        }
        foreach (var b in gm.buildings)
        {
            if (b == null) continue;
            data.buildings.Add(new SaveSystem.BuildingSnap {
                type = (int)b.type, teamId = b.teamId,
                x = b.transform.position.x, z = b.transform.position.z, hp = b.maxHp,
            });
        }
        if (gm.triggers != null) data.triggers = gm.triggers.Snapshot();

        PlayerPrefs.SetString(ScenarioKey, JsonUtility.ToJson(data));
        PlayerPrefs.Save();
        SetStatus($"Senaryo kaydedildi ({data.units.Count}b+{data.buildings.Count}bi).");
    }

    void LoadScenario()
    {
        string json = PlayerPrefs.GetString(ScenarioKey, "");
        if (string.IsNullOrEmpty(json)) { SetStatus("Kayıtlı senaryo bulunamadı."); return; }
        var data = JsonUtility.FromJson<SaveSystem.SaveData>(json);
        if (data == null) { SetStatus("Senaryo okunamadı."); return; }
        GameBootstrap.PendingLoad = data;
        Close();
        GameBootstrap.Restart(data.mapSeed);
    }

    void PlaytestScenario()
    {
        SaveScenario(); // snapshot current state
        string json = PlayerPrefs.GetString(ScenarioKey, "");
        var data = JsonUtility.FromJson<SaveSystem.SaveData>(json);
        if (data == null) return;
        GameBootstrap.PendingLoad = data;
        Close();
        GameBootstrap.Restart(data.mapSeed);
    }

    // ── UI ────────────────────────────────────────────────────────────────────

    void BuildUI()
    {
        if (_canvas != null) return; // already built

        if (Object.FindAnyObjectByType<UnityEngine.EventSystems.EventSystem>() == null)
        {
            var es = new GameObject("EditorEventSystem");
            es.AddComponent<UnityEngine.EventSystems.EventSystem>();
            es.AddComponent<UnityEngine.EventSystems.StandaloneInputModule>();
        }

        var cgo = new GameObject("ScenarioEditorCanvas");
        cgo.transform.SetParent(transform, false);
        _canvas = cgo.AddComponent<Canvas>();
        _canvas.renderMode   = RenderMode.ScreenSpaceOverlay;
        _canvas.sortingOrder = 6000;
        var scaler = cgo.AddComponent<CanvasScaler>();
        scaler.uiScaleMode         = CanvasScaler.ScaleMode.ScaleWithScreenSize;
        scaler.referenceResolution = new Vector2(1920, 1080);
        cgo.AddComponent<GraphicRaycaster>();

        // ── Top bar ────────────────────────────────────────────────────────────
        var bar = MakeRect("TopBar", cgo.transform);
        bar.anchorMin = new Vector2(0,1); bar.anchorMax = new Vector2(1,1);
        bar.pivot     = new Vector2(0.5f, 1f);
        bar.sizeDelta = new Vector2(0, 44);
        bar.anchoredPosition = Vector2.zero;
        bar.gameObject.AddComponent<Image>().color = PanelBg;

        MakeLabel(bar, "📝 Senaryo Editörü", -700, 0, 22, TextCol, FontStyle.Bold);

        _statusText = MakeLabel(bar, "", 0, 0, 16, LabelCol, FontStyle.Normal);

        MakeBtn(bar, "🗑️ Sil", 550, 0, 90, 36, () =>
        {
            _deleteMode = !_deleteMode;
            _selectedItem = -1;
            SetStatus(_deleteMode ? "Silme modu — haritada tıkla." : "Silme modu kapatıldı.");
        });
        MakeBtn(bar, "💾 Kaydet", 660, 0, 100, 36, SaveScenario);
        MakeBtn(bar, "📂 Yükle",  770, 0, 100, 36, LoadScenario);
        MakeBtn(bar, "▶ Test Et", 880, 0, 110, 36, PlaytestScenario);
        MakeBtn(bar, "✕ Kapat",   900, 0,  90, 36, Close);

        // ── Left palette panel ─────────────────────────────────────────────────
        var pal = MakeRect("Palette", cgo.transform);
        pal.anchorMin = new Vector2(0, 0); pal.anchorMax = new Vector2(0, 1);
        pal.pivot = new Vector2(0, 0.5f);
        pal.sizeDelta = new Vector2(180, -44);
        pal.anchoredPosition = new Vector2(0, -22);
        pal.gameObject.AddComponent<Image>().color = PanelBg;

        // Category tabs
        for (int i = 0; i < Categories.Length; i++)
        {
            int captured = i;
            MakeBtn(pal, Categories[i], 0, 440 - i * 36, 170, 32, () =>
            {
                _selectedCategory = captured;
                _selectedItem     = -1;
                RefreshPalette(pal);
            });
        }

        // Team selector
        MakeLabel(pal, "Takım:", 0, 310, 16, LabelCol, FontStyle.Normal);
        for (int t = 0; t < 4; t++)
        {
            int captured = t;
            var col = TeamPalette.For(t);
            var tb = MakeBtn(pal, $"T{t}", -55 + t * 40, 280, 34, 28, () => { _selectedTeam = captured; });
            tb.GetComponent<Image>().color = new Color(col.r * 0.6f, col.g * 0.6f, col.b * 0.6f, 0.9f);
        }

        // Entity list placeholder (filled in RefreshPalette)
        RefreshPalette(pal);

        // ── Right panel: player setup ──────────────────────────────────────────
        var rp = MakeRect("PlayerSetup", cgo.transform);
        rp.anchorMin = new Vector2(1, 0); rp.anchorMax = new Vector2(1, 1);
        rp.pivot = new Vector2(1, 0.5f);
        rp.sizeDelta = new Vector2(200, -44);
        rp.anchoredPosition = new Vector2(0, -22);
        rp.gameObject.AddComponent<Image>().color = PanelBg;
        BuildPlayerPanel(rp);
    }

    void RefreshPalette(RectTransform pal)
    {
        _paletteButtons.Clear();
        // Remove old item buttons (keep first 3 category + 5 team buttons = first 8 children)
        for (int i = pal.childCount - 1; i >= 8; i--)
            Object.Destroy(pal.GetChild(i).gameObject);

        string[] names = _selectedCategory == 0 ? GetNames(UnitItems.Length, 0)
                       : _selectedCategory == 1 ? GetNames(BuildingItems.Length, 1)
                       :                          GetNames(ResourceItems.Length, 2);

        for (int i = 0; i < names.Length; i++)
        {
            int captured = i;
            var btn = MakeBtn(pal, names[i], 0, 240 - i * 34, 170, 30, () =>
            {
                _selectedItem = captured;
                _deleteMode   = false;
                SetStatus($"Seçildi: {GetItemName(_selectedCategory, _selectedItem)} — haritaya tıkla.");
            });
            _paletteButtons.Add((btn.GetComponent<Button>(), i));
        }
    }

    string[] GetNames(int count, int cat)
    {
        var n = new string[count];
        for (int i = 0; i < count; i++)
            n[i] = GetItemName(cat, i);
        return n;
    }

    string GetItemName(int cat, int idx) => cat switch
    {
        0 when idx < UnitItems.Length     => UnitItems[idx].label,
        1 when idx < BuildingItems.Length => BuildingItems[idx].label,
        2 when idx < ResourceItems.Length => ResourceItems[idx].label,
        _ => "?"
    };

    void BuildPlayerPanel(RectTransform rp)
    {
        MakeLabel(rp, "Oyuncu Ayarları", 0, 480, 18, TextCol, FontStyle.Bold);
        var gm = GameManager.Instance;
        if (gm == null) return;

        for (int t = 0; t < 4; t++)
        {
            int ti = t;
            float y = 420 - t * 90;
            var col = TeamPalette.For(t);
            MakeLabel(rp, $"Takım {t}", -60, y + 35, 16,
                new Color(col.r, col.g, col.b), FontStyle.Bold);

            // Food +/-
            MakeLabel(rp, $"Yiyecek:{gm.teamRes[t].food}", 0, y + 14, 14, LabelCol, FontStyle.Normal);
            MakeBtn(rp, "+500", 60, y, 60, 24, () =>
            {
                gm.teamRes[ti].food += 500;
                Object.Destroy(rp.gameObject); // rebuild panel
                var parent = rp.parent;
                var newPanel = MakeRect("PlayerSetup", parent);
                newPanel.anchorMin = new Vector2(1,0); newPanel.anchorMax = new Vector2(1,1);
                newPanel.pivot = new Vector2(1, 0.5f);
                newPanel.sizeDelta = new Vector2(200, -44);
                newPanel.anchoredPosition = new Vector2(0, -22);
                newPanel.gameObject.AddComponent<Image>().color = PanelBg;
                BuildPlayerPanel(newPanel);
            });

            // Age up
            MakeBtn(rp, "Çağ+", -60, y - 20, 80, 24, () =>
            {
                var tech = gm.teamTech[ti];
                if (tech.age < Age.Imperial)
                {
                    var next = (Age)((int)tech.age + 1);
                    ResearchSystem.Apply(next == Age.Feudal   ? TechType.FeudalAge
                                     :  next == Age.Castle   ? TechType.CastleAge
                                     :                         TechType.ImperialAge, ti);
                }
            });
        }
    }

    void SetStatus(string msg) { if (_statusText != null) _statusText.text = msg; }

    // ── UI helpers ────────────────────────────────────────────────────────────

    static RectTransform MakeRect(string name, Transform parent)
    {
        var go = new GameObject(name);
        go.transform.SetParent(parent, false);
        return go.AddComponent<RectTransform>();
    }

    static Text MakeLabel(RectTransform parent, string text, float x, float y,
        int fontSize, Color col, FontStyle style)
    {
        var go = new GameObject("Lbl");
        go.transform.SetParent(parent, false);
        var rt = go.AddComponent<RectTransform>();
        rt.sizeDelta = new Vector2(180, 26);
        rt.anchoredPosition = new Vector2(x, y);
        var t = go.AddComponent<Text>();
        t.text      = text;
        t.fontSize  = fontSize;
        t.color     = col;
        t.fontStyle = style;
        t.alignment = TextAnchor.MiddleCenter;
        t.font      = null;
        return t;
    }

    static GameObject MakeBtn(RectTransform parent, string label, float x, float y,
        float w, float h, System.Action onClick)
    {
        var go = new GameObject("Btn");
        go.transform.SetParent(parent, false);
        var rt = go.AddComponent<RectTransform>();
        rt.sizeDelta = new Vector2(w, h);
        rt.anchoredPosition = new Vector2(x, y);
        var img = go.AddComponent<Image>();
        img.color = BtnColor;
        var btn = go.AddComponent<Button>();
        var cols = btn.colors;
        cols.normalColor      = Color.white;
        cols.highlightedColor = new Color(1.3f, 1.3f, 1.3f);
        cols.pressedColor     = new Color(1.6f, 1.6f, 1.6f);
        btn.colors = cols;
        btn.onClick.AddListener(() => onClick());
        MakeLabel(rt, label, 0, 0, 13, TextCol, FontStyle.Normal);
        return go;
    }
}

File: web/src/game/ScenarioEditor.ts
/**
 * ScenarioEditor.ts — Runtime map editor (N12.edit port, simplified).
 * Toggle with 'E' key. Left palette → click terrain → entity placed.
 * Delete mode: toggle 🗑 → click entity → removed.
 * Save/load via localStorage (JSON).
 */
import * as THREE from "three";
import { BuildingType, ResourceKind, UnitType } from "../core/GameTypes";
import type { Unit } from "./Unit";
import type { Building } from "./Building";
import type { ResourceNode } from "./ResourceNode";

const SAVE_KEY = "AoA_Scenario_0";

type Placeable =
  | { kind: 'unit'; type: UnitType; label: string }
  | { kind: 'building'; type: BuildingType; label: string }
  | { kind: 'resource'; rtype: ResourceKind; amount: number; label: string };

const PALETTE: Placeable[] = [
  { kind: 'unit', type: UnitType.Villager,  label: 'Köylü' },
  { kind: 'unit', type: UnitType.Militia,   label: 'Nefer' },
  { kind: 'unit', type: UnitType.Archer,    label: 'Okçu' },
  { kind: 'unit', type: UnitType.Cavalry,   label: 'Süvari' },
  { kind: 'unit', type: UnitType.Trebuchet, label: 'Trebuchet' },
  { kind: 'building', type: BuildingType.TownCenter, label: 'Şehir M.' },
  { kind: 'building', type: BuildingType.Barracks,   label: 'Kışla' },
  { kind: 'building', type: BuildingType.WatchTower, label: 'Kule' },
  { kind: 'building', type: BuildingType.Castle,     label: 'Kale' },
  { kind: 'resource', rtype: ResourceKind.Gold,  amount: 800, label: 'Altın' },
  { kind: 'resource', rtype: ResourceKind.Wood,  amount: 300, label: 'Odun' },
  { kind: 'resource', rtype: ResourceKind.Stone, amount: 400, label: 'Taş' },
  { kind: 'resource', rtype: ResourceKind.Food,  amount: 250, label: 'Yiyecek' },
];

interface Callbacks {
  spawnUnit:     (type: UnitType, x: number, z: number, team: number) => Unit | null;
  placeBuilding: (type: BuildingType, x: number, z: number, team: number) => void;
  placeResource: (kind: ResourceKind, amount: number, x: number, z: number) => void;
  removeUnit:     (u: Unit) => void;
  removeBuilding: (b: Building) => void;
  removeResource: (n: ResourceNode) => void;
  getUnits:       () => Unit[];
  getBuildings:   () => Building[];
  getResources:   () => ResourceNode[];
}

export class ScenarioEditor {
  private _open       = false;
  private _deleteMode = false;
  private _selTeam    = 0;
  private _selItem    = -1;
  private readonly _panel: HTMLDivElement;
  private readonly _raycaster = new THREE.Raycaster();
  private readonly _ground    = new THREE.Plane(new THREE.Vector3(0, 1, 0), 0);
  private readonly _hitPt     = new THREE.Vector3();
  private _activeBtn: HTMLButtonElement | null = null;
  private readonly _callbacks: Callbacks;
  private readonly _camera: THREE.Camera;
  private readonly _canvas: HTMLCanvasElement;

  constructor(
    container: HTMLElement,
    camera: THREE.Camera,
    canvas: HTMLCanvasElement,
    callbacks: Callbacks,
  ) {
    this._camera    = camera;
    this._canvas    = canvas;
    this._callbacks = callbacks;

    this._panel = document.createElement("div");
    this._panel.style.cssText = `
      position:absolute; top:10px; left:10px; width:160px;
      background:rgba(6,10,20,0.92); border:1px solid #334; border-radius:8px;
      padding:10px 8px; font-family:monospace; font-size:12px; color:#c8c8e0;
      display:none; z-index:800; user-select:none;
    `;
    container.appendChild(this._panel);

    canvas.addEventListener("click", (e) => this._onClick(e));
  }

  isOpen(): boolean { return this._open; }

  toggle(): void {
    this._open = !this._open;
    this._panel.style.display = this._open ? "block" : "none";
    if (this._open) this._rebuild();
  }

  private _rebuild(): void {
    this._panel.innerHTML = "";

    const title = document.createElement("div");
    title.textContent = "SENARYO EDITÖRÜ";
    title.style.cssText = "font-size:10px;color:#f5d060;font-weight:bold;letter-spacing:1px;margin-bottom:8px;";
    this._panel.appendChild(title);

    // Team selector
    const teamRow = document.createElement("div");
    teamRow.style.cssText = "display:flex;gap:4px;margin-bottom:8px;";
    for (let t = 0; t < 4; t++) {
      const btn = document.createElement("button");
      btn.textContent = `T${t}`;
      btn.style.cssText = `
        flex:1; padding:3px; border:1px solid ${t === this._selTeam ? '#f5d060' : '#333'};
        border-radius:3px; background:${t === this._selTeam ? '#2a2000' : '#11192a'};
        color:${t === this._selTeam ? '#f5d060' : '#aaa'}; font-size:10px; cursor:pointer;
        font-family:monospace;
      `;
      btn.addEventListener("click", () => { this._selTeam = t; this._rebuild(); });
      teamRow.appendChild(btn);
    }
    this._panel.appendChild(teamRow);

    // Delete mode toggle
    const delBtn = document.createElement("button");
    delBtn.textContent = this._deleteMode ? "SİL MODU AÇIK" : "Sil Modu";
    delBtn.style.cssText = `
      width:100%; padding:4px; margin-bottom:8px; border:1px solid ${this._deleteMode ? '#cc4444' : '#333'};
      border-radius:3px; background:${this._deleteMode ? '#3a0a0a' : '#11192a'};
      color:${this._deleteMode ? '#ff6666' : '#aaa'}; font-size:11px; cursor:pointer;
      font-family:monospace;
    `;
    delBtn.addEventListener("click", () => { this._deleteMode = !this._deleteMode; this._selItem = -1; this._rebuild(); });
    this._panel.appendChild(delBtn);

    if (!this._deleteMode) {
      const sep = document.createElement("div");
      sep.textContent = "── PALET ──";
      sep.style.cssText = "font-size:10px;color:#556;margin-bottom:6px;text-align:center;";
      this._panel.appendChild(sep);

      // Palette buttons
      PALETTE.forEach((p, i) => {
        const btn = document.createElement("button");
        btn.textContent = p.label;
        const sel = i === this._selItem;
        btn.style.cssText = `
          width:100%; padding:4px 6px; margin-bottom:3px;
          border:1px solid ${sel ? '#6af' : '#333'}; border-radius:3px;
          background:${sel ? '#0a1a2a' : '#11192a'}; color:${sel ? '#6af' : '#aaa'};
          font-size:11px; cursor:pointer; text-align:left; font-family:monospace;
        `;
        btn.addEventListener("click", () => {
          this._selItem = i;
          this._rebuild();
        });
        if (sel) this._activeBtn = btn;
        this._panel.appendChild(btn);
      });
    }

    // Save/Load
    const ioRow = document.createElement("div");
    ioRow.style.cssText = "display:flex;gap:4px;margin-top:10px;";

    const saveBtn = document.createElement("button");
    saveBtn.textContent = "Kaydet";
    saveBtn.style.cssText = `
      flex:1; padding:4px; border:1px solid #336; border-radius:3px;
      background:#0a0e1e; color:#88aadd; font-size:10px; cursor:pointer; font-family:monospace;
    `;
    saveBtn.addEventListener("click", () => this._save());
    ioRow.appendChild(saveBtn);

    const loadBtn = document.createElement("button");
    loadBtn.textContent = "Yükle";
    loadBtn.style.cssText = `
      flex:1; padding:4px; border:1px solid #336; border-radius:3px;
      background:#0a0e1e; color:#88aadd; font-size:10px; cursor:pointer; font-family:monospace;
    `;
    loadBtn.addEventListener("click", () => this._load());
    ioRow.appendChild(loadBtn);

    this._panel.appendChild(ioRow);

    const hint = document.createElement("div");
    hint.textContent = "E: kapat";
    hint.style.cssText = "font-size:9px;color:#445;margin-top:8px;text-align:center;";
    this._panel.appendChild(hint);
  }

  private _onClick(e: MouseEvent): void {
    if (!this._open) return;
    // Ignore clicks inside the editor panel itself
    if (this._panel.contains(e.target as Node)) return;

    const rect = this._canvas.getBoundingClientRect();
    const ndc = new THREE.Vector2(
      ((e.clientX - rect.left) / rect.width) * 2 - 1,
      -((e.clientY - rect.top)  / rect.height) * 2 + 1,
    );
    this._raycaster.setFromCamera(ndc, this._camera);
    if (!this._raycaster.ray.intersectPlane(this._ground, this._hitPt)) return;

    const x = this._hitPt.x;
    const z = this._hitPt.z;

    if (this._deleteMode) {
      this._tryDelete(x, z);
      return;
    }

    if (this._selItem < 0 || this._selItem >= PALETTE.length) return;
    const p = PALETTE[this._selItem];

    if (p.kind === 'unit') {
      this._callbacks.spawnUnit(p.type, x, z, this._selTeam);
    } else if (p.kind === 'building') {
      this._callbacks.placeBuilding(p.type, x, z, this._selTeam);
    } else if (p.kind === 'resource') {
      this._callbacks.placeResource(p.rtype, p.amount, x, z);
    }
  }

  private _tryDelete(x: number, z: number): void {
    const RADIUS = 3;
    const r2 = RADIUS * RADIUS;

    for (const u of this._callbacks.getUnits()) {
      if (!u.alive) continue;
      const dx = u.x - x; const dz = u.z - z;
      if (dx * dx + dz * dz <= r2) { this._callbacks.removeUnit(u); return; }
    }
    for (const b of this._callbacks.getBuildings()) {
      if (!b.alive) continue;
      const dx = b.pos.x - x; const dz = b.pos.z - z;
      if (dx * dx + dz * dz <= r2) { this._callbacks.removeBuilding(b); return; }
    }
    for (const n of this._callbacks.getResources()) {
      if (n.depleted) continue;
      const dx = n.root.position.x - x; const dz = n.root.position.z - z;
      if (dx * dx + dz * dz <= r2) { this._callbacks.removeResource(n); return; }
    }
  }

  private _save(): void {
    const data = {
      units:     this._callbacks.getUnits().filter(u => u.alive).map(u => ({ type: u.unitType, x: u.x, z: u.z, team: u.teamId })),
      buildings: this._callbacks.getBuildings().filter(b => b.alive).map(b => ({ type: b.buildingType, x: b.pos.x, z: b.pos.z, team: b.teamId })),
      resources: this._callbacks.getResources().filter(n => !n.depleted).map(n => ({ rtype: n.kind, amount: n.amount, x: n.root.position.x, z: n.root.position.z })),
    };
    localStorage.setItem(SAVE_KEY, JSON.stringify(data));
    console.info("[ScenarioEditor] kaydedildi");
  }

  private _load(): void {
    const raw = localStorage.getItem(SAVE_KEY);
    if (!raw) { console.warn("[ScenarioEditor] kayıtlı senaryo yok"); return; }
    try {
      const data = JSON.parse(raw) as {
        units:     Array<{type: UnitType; x: number; z: number; team: number}>;
        buildings: Array<{type: BuildingType; x: number; z: number; team: number}>;
        resources: Array<{rtype: ResourceKind; amount: number; x: number; z: number}>;
      };
      for (const u of data.units ?? []) this._callbacks.spawnUnit(u.type, u.x, u.z, u.team);
      for (const b of data.buildings ?? []) this._callbacks.placeBuilding(b.type, b.x, b.z, b.team);
      for (const r of data.resources ?? []) this._callbacks.placeResource(r.rtype, r.amount, r.x, r.z);
      console.info("[ScenarioEditor] yüklendi");
    } catch { console.error("[ScenarioEditor] JSON parse hatası"); }
  }
}
